x-pack/legacy/plugins/siem/public/pages/detection_engine/components/signals/helpers.ts

```diff
@@ -0,0 +1,137 @@
+  'source.ip',
+  'network.community_id',
+  'user.name',
+  'process.name',
+];
+
+export const findValueToChangeInQuery = (
+  keuryNode: esKuery.KueryNode,
+  valueToChange: FindValueToChangeInQuery[] = []
+): FindValueToChangeInQuery[] => {
+  let localValueToChange = valueToChange;
+  if (keuryNode.function === 'is' && templateFields.includes(keuryNode.arguments[0].value)) {
+    localValueToChange = [
+      ...localValueToChange,
+      {
+        field: keuryNode.arguments[0].value,
+        valueToChange: keuryNode.arguments[1].value,
+      },
+    ];
+  }
+  return keuryNode.arguments.reduce(
+    (addValueToChange: FindValueToChangeInQuery[], ast: esKuery.KueryNode) => {
+      if (ast.function === 'is' && templateFields.includes(ast.arguments[0].value)) {
+        return [
+          ...addValueToChange,
+          {
+            field: ast.arguments[0].value,
+            valueToChange: ast.arguments[1].value,
+          },
+        ];
+      }
+      if (ast.arguments) {
+        return findValueToChangeInQuery(ast, addValueToChange);
+      }
+      return addValueToChange;
+    },
+    localValueToChange
+  );
+};
+
+export const replaceTemplateFieldFromQuery = (query: string, ecsData: Ecs) => {
+  if (query !== '') {
+    const valueToChange = findValueToChangeInQuery(esKuery.fromKueryExpression(query));
+    return valueToChange.reduce((newQuery, vtc) => {
+      const newValue = get(vtc.field, ecsData);
+      if (newValue != null) {
+        return newQuery.replace(vtc.valueToChange, newValue);
+      }
+      return newQuery;
+    }, query);
+  }
+  return '';
+};
+
+export const replaceTemplateFieldFromMatchFilters = (filters: esFilters.Filter[], ecsData: Ecs) => {
+  if (!isEmpty(filters)) {
+    return filters.map(filter => {
+      if (
+        filter.meta.type === 'phrase' &&
+        filter.meta.key != null &&
+        templateFields.includes(filter.meta.key)
+      ) {
+        const newValue = get(filter.meta.key, ecsData);
+        if (newValue != null) {
+          filter.meta.params = { query: newValue };
+          filter.query = { match_phrase: { [filter.meta.key]: newValue } };
+        }
+      }
+      return filter;
+    });
+  }
+  return [];
+};
+
+export const reformatDataProviderWithNewValue = <T extends DataProvider | DataProvidersAnd>(
+  dataProvider: T,
+  ecsData: Ecs
+): T => {
+  if (templateFields.includes(dataProvider.queryMatch.field)) {
+    const newValue = get(dataProvider.queryMatch.field, ecsData);
+    if (newValue != null) {
+      dataProvider.id = dataProvider.id.replace(dataProvider.name, newValue);
+      dataProvider.name = newValue;
+      dataProvider.queryMatch.value = newValue;
+      dataProvider.queryMatch.displayField = undefined;
+      dataProvider.queryMatch.displayValue = undefined;
+    }
+  }
+  return dataProvider;
+};
+
+export const replaceTemplateFieldFromDataProviders = (
+  dataProviders: DataProvider[],
+  ecsData: Ecs
+) => {
+  if (!isEmpty(dataProviders)) {
```

Comment:
Same comment, but just an extra bit which is utilizing `isEmpty` begins including regression bugs where if someone accidentally changes code to send in a null or undefined we won't ever see it. After a while maintainers begin to get nervous about removing additional `isEmpty` checks which is why I'm not really for them.

If the types say it can be an array then it should be and we can just call regular `map` , filter, etc... on them and not worry about our runtime types drifting too far from our compile time types.